velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,195 @@
+    }
+
+    const char* const begin = current.begin();
+    const char* const end = current.end();
+    const char* pos = begin;
+    while (pos < end && pos < limit + begin) {
+      arrayWriter.add_item().setNoCopy(StringView(pos, 1));
+      pos += 1;
+    }
+    resultWriter.commit();
+  }
 
-  void apply(
-      const SelectivityVector& rows,
-      std::vector<VectorPtr>& args,
-      const TypePtr& /* outputType */,
-      exec::EvalCtx& context,
-      VectorPtr& result) const override {
-    auto delimiterVector = args[1]->as<ConstantVector<StringView>>();
-    VELOX_CHECK(
-        delimiterVector, "Split function supports only constant delimiter");
-    auto patternString = args[1]->as<ConstantVector<StringView>>()->valueAt(0);
-    VELOX_CHECK_EQ(
-        patternString.size(),
-        1,
-        "split only supports only single-character pattern");
-    char pattern = patternString.data()[0];
-    SplitCharacter splitCharacter(pattern);
-    splitCharacter.apply(rows, args, nullptr, context, result);
+  // Split with a non-empty pattern.
+  void splitInner(
+      StringView input,
+      const StringView delim,
+      int64_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    // Add new array (for the new row) to our array vector.
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+
+    // Trivial case of converting string to array with 1 element.
+    if (limit == 1) {
+      arrayWriter.add_item().setNoCopy(input);
+      resultWriter.commit();
+      return;
+    }
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // adding them to the elements vector, until we reached the end of the
+    // string or the limit.
+    int32_t addedElements{0};
+    bool emptyDelim = delim.size() == 0 ? true : false;
+    auto* re = cache_.findOrCompile(delim);
+    const auto re2String = re2::StringPiece(input.data(), input.size());
+    size_t pos = 0;
+    const char* start = input.data();
+    re2::StringPiece subMatches[1];
+    while (re->Match(
+        re2String, pos, input.size(), RE2::Anchor::UNANCHORED, subMatches, 1)) {
+      const auto fullMatch = subMatches[0];
+      auto offset = fullMatch.data() - start;
+      const auto size = fullMatch.size();
+
+      if (offset >= input.size()) {
+        break;
+      }
+
+      arrayWriter.add_item().setNoCopy(
+          StringView(input.data() + pos, offset - pos));
+      pos = offset + size;
+      ++addedElements;
+      // If the next element should be the last, leave the loop.
+      if (addedElements + 1 == limit) {
+        break;
+      }
+    }
+
+    // Add the rest of the string and we are done.
+    // Note, that the rest of the string can be empty - we still add it.
+    arrayWriter.add_item().setNoCopy(
+        StringView(input.data() + pos, input.size() - pos));
+    resultWriter.commit();
   }
+
+  mutable functions::detail::ReCache cache_;
```

Comment:
nit: add comments for this variable.
I wonder why we need the keyword mutable. It allows us to modify this variable from const functions, but in current implementation cache_  has not been modified. Am I miss something?